File: frontend/src/api/hierarchy.ts
import { apiFetch } from "./client";

export interface HierarchyNode {
  id: string;
  kind: string;
  label: string;
  template_label?: string | null;
  children: HierarchyNode[];
}

export interface VehicleHierarchy {
  vehicle_id: string;
  revision_id: string;
  root: HierarchyNode;
}

export function fetchHierarchy(vehicleId: string, revisionId: string) {
  return apiFetch<VehicleHierarchy>(
    `/vehicles/${vehicleId}/revisions/${revisionId}/hierarchy`,
  );
}

File: backend/app/schemas/hierarchy.py
from uuid import UUID

from pydantic import BaseModel


class HierarchyNode(BaseModel):
    id: UUID
    kind: str
    label: str
    template_label: str | None = None
    children: list["HierarchyNode"] = []


HierarchyNode.model_rebuild()


class VehicleHierarchyResponse(BaseModel):
    vehicle_id: UUID
    revision_id: UUID
    root: HierarchyNode
